recirq/qml_lfe/sample_analysis.ipynb

```diff
@@ -0,0 +1,1451 @@
+     "output_type": "stream",
+     "text": [
+      "ACC 1.0\n"
+     ]
+    },
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "from tqdm import tqdm\n",
+    "import cirq\n",
+    "import numpy as np\n",
+    "from os import listdir\n",
+    "import random\n",
+    "from os.path import isfile, join\n",
+    "from cirq import Simulator\n",
+    "simulator = Simulator()\n",
+    "\n",
+    "all_paulis = dict([(\"I\",0), (\"X\",1), (\"Y\",2), (\"Z\",3)])\n",
+    "\n",
+    "def entire_process_predict_Pauli_observable(qubits, hidden_P, num_rep):\n",
+    "    n = len(qubits) // 2\n",
+    "    output = []\n",
+    "    \n",
+    "    sign_p = -1 if np.random.random() < 0.5 else 1\n",
+    "    for _ in range(num_rep):\n",
+    "        circuit = cirq.Circuit()\n",
+    "\n",
+    "        last_i = 0\n",
+    "        for i, pauli in enumerate(hidden_P):\n",
+    "            if pauli != \"I\":\n",
+    "                last_i = i\n",
+    "            \n",
+    "        for twocopy in [0, 1]:\n",
+    "            parity = sign_p\n",
+    "            for i, pauli in enumerate(hidden_P):\n",
+    "                if pauli != \"I\":\n",
+    "                    if last_i == i:\n",
+    "                        if parity == -1:\n",
+    "                            circuit += cirq.X(qubits[2*i+twocopy])\n",
+    "                    elif random.choice([0, 1]) == 1:\n",
+    "                        parity *= -1\n",
+    "                        circuit += cirq.X(qubits[2*i+twocopy])\n",
+    "                else:\n",
+    "                    if random.choice([0, 1]) == 1:\n",
+    "                        circuit += cirq.X(qubits[2*i+twocopy])\n",
+    "\n",
+    "                if pauli == \"X\":\n",
+    "                    circuit += cirq.H(qubits[2*i+twocopy])\n",
+    "                if pauli == \"Y\":\n",
+    "                    circuit += cirq.H(qubits[2*i+twocopy])\n",
+    "                    circuit += cirq.S(qubits[2*i+twocopy])\n",
+    "\n",
+    "        for i, pauli in enumerate(hidden_P):\n",
+    "            circuit += cirq.CNOT(qubits[2*i], qubits[2*i+1])\n",
+    "            circuit += cirq.H(qubits[2*i])\n",
+    "\n",
+    "        for i, qubit in enumerate(qubits):\n",
+    "            circuit += cirq.measure(qubit, key='q{}'.format(i))\n",
+    "\n",
+    "        samples = simulator.run(circuit, repetitions=1)\n",
+    "\n",
+    "        raw_samples = np.zeros(len(qubits), dtype=int)\n",
+    "        for i in range(len(qubits)):\n",
+    "            raw_samples[i] = samples.data['q{}'.format(i)]\n",
+    "\n",
+    "        output.append(raw_samples)\n",
+    "\n",
+    "    return output\n",
+    "\n",
+    "def sample_one(b_experi, n_qubits):\n",
+    "    qubits = cirq.GridQubit.rect(n_qubits, 2)\n",
+    "\n",
+    "    true_p = [random.choice([\"I\", \"X\", \"Y\", \"Z\"]) for i in range(n_qubits)]\n",
```

Comment:
Optional nit: Should `random.choice(["I", "X", "Y", "Z"]) for i in range(n_qubits)` be made into a function?  It's used in quite a few places here.